Add autocompletion for custom emoji

autumnontape opened the pull request:
Closes #398.

Maybe it'd be a good idea to add more tests?

I don't have much Android experience, so I might not be doing everything in the best way.

![autocompletion of "hello :masto", showing ":masto_ball:" and ":mastodon:" as suggestions, with high-resolution previews of the emoji in question](https://user-images.githubusercontent.com/40726037/53538030-e9abe800-3ac0-11e9-959b-d861ea918765.png)

(Don't worry about the low resolution of the ":mastodon:" emoji, that's just what it looks like on that instance.)

autumnontape commented:
Heads up, I'm going to do a force push in a few minutes to get it down to two commits.

autumnontape commented:
It may be better to move the `emojiListRetrievalLatch.countDown()` calls into `setEmojiList()` and either remove the `emojiList` argument to `setEmojiList()` or have `setEmojiList()` set the `emojiList` field itself rather than relying on the field having been set already. There doesn't seem to be any reason not to make any of these changes.

autumnontape commented:
I went ahead and made the last change I mentioned in the former comment. I think this makes the code less prone to future error.

charlag commented on app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java:
```diff
@@ -1569,6 +1571,28 @@ private void saveDraftAndFinish() {
                     }
+                case ':':
+                    try {
+                        emojiListRetrievalLatch.await();
```
I'm not sure we need to wait for latch. Autocomplete should be fast and if it doesn't work then maybe it makes sense to just skip. Did you notice the change when you used it?

autumnontape replied:
The latch is there so that if you begin to type an emoji code and then stop, the result will always be consistent. So you'll never have to type an additional character and then delete it to see the results you want, if you're using autocomplete as a search.

The case where the `await` doesn't return immediately is out of the ordinary, and I have a fast connection, so I don't notice it in practice. I did test the feature with a throttled connection, and it works as expected.

I could remove the latch without affecting most potential users. I just figure it's a pretty small difference in code for some nice consistency.

charlag replied:
Yeah, maybe it's for the best there

charlag commented on app/src/main/res/layout/item_autocomplete_emoji.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Please change `RelativeLayout` to be something else please (`LinearLayout` or `ConstraintLayout`). Ask for help if needed.

Tak commented on app/src/main/java/com/keylesspalace/tusky/adapter/MentionTagAutoCompleteAdapter.java:
```diff
@@ -44,6 +45,7 @@
         implements Filterable {
     private static final int ACCOUNT_VIEW_TYPE = 0;
     private static final int HASHTAG_VIEW_TYPE = 1;
+    private static final int EMOJI_VIEW_TYPE = 2;
```
Maybe it makes sense to convert this to a Real Enum™ at this point?

autumnontape replied:
Sure, I'll do that real quick

autumnontape replied:
Oh, `BaseAdapter` mandates that a lot of the methods that deal with these constants use ints. It probably wouldn't be worthwhile to use an enum, then.

Tak commented:
Maybe it makes sense to show the completion results without the colons?
:+1: overall

charlag commented:
I think colons are okay too

autumnontape commented:
The emoji shortcodes feel naked to me without colons :stuck_out_tongue:

autumnontape commented:
All right, I converted the RelativeLayout to a LinearLayout as @charlag asked me to. Android Studio is telling me now that I could replace the whole thing with just a TextView with compound drawables, but doing that doesn't seem worth it, since it doesn't look like it'd be straightforward to hook a compound drawable into Picasso.

autumnontape commented:
Hmm, hang on, I accidentally decreased the padding between the emoji preview and the shortcode. I liked it better the way it was before, so I'm going to fix that and force-push again.

charlag commented:
Huh, it can actually be a TextView, that's right. It's not hard to hook it to Picasso either, we have already some places where we use `into(new Target())` (maybe in Kotlin), you can check it

autumnontape commented:
Okay, the LinearLayout is pixel-for-pixel identical to the old RelativeLayout now.

autumnontape commented:
@charlag I'll check it out.

charlag commented:
I don't mind either, just saying

autumnontape commented:
It looks like I'd have to add code to set the size of the image and code to make it work with Picasso, if I used a compound drawable here. It doesn't really seem like a step forward.

Any more feedback?

charlag commented:
I wouldn't bother tbh, let's wait until @connyduck looks at this
Great job btw!

autumnontape commented:
Thanks!

autumnontape commented:
Hmm, maybe `MentionTagAutoCompleteAdapter` should be renamed to `ComposeAutoCompleteAdapter` or something.

charlag commented:
That's a good idea

autumnontape commented:
All right, I've renamed `MentionTagAutoCompleteAdapter`, `MentionTagTokenizer`, and `MentionTagTokenizerTest` to start with "Compose" instead.

connyduck commented on app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java:
```diff
@@ -1557,18 +1559,40 @@ private void saveDraftAndFinish() {
+
+                        List<Emoji> results = CollectionsKt.filter(
+                                emojiList,
+                                emoji -> emoji.getShortcode().startsWith(incomplete)
```
Why only `startsWith`? Wouldn't it be better to return all emojis that contain the string, with emojis that start with it first?

autumnontape replied:
Wow, that's a great idea! The second list will be a lot longer than the first in many cases, so maybe it'd be a good idea to put a horizontal rule or something between them?

I should probably also make the matching case-insensitive.

connyduck commented on app/src/main/res/layout/item_autocomplete_emoji.xml:
```diff
@@ -0,0 +1,33 @@
+        android:maxLines="1"
+        android:textColor="?android:textColorPrimary"
+        android:textSize="?attr/status_text_large"
+        android:textStyle="normal|bold"
```
Bold? It looks very different than the autocomplete of a hashtag, I think thats weird. But I actually like the bold, maybe add a dedicated autocomplete layout for hashtags that is bold as well?

autumnontape replied:
All right, I'll try it.

autumnontape replied:
I wasn't sure how it'd be, but I like the bold hashtag autocomplete a lot. Now all three kinds of autocomplete bold their results. :+1:

connyduck commented on app/src/main/java/com/keylesspalace/tusky/adapter/ComposeAutoCompleteAdapter.java:
```diff
@@ -159,6 +163,31 @@ public View getView(int position, @Nullable View convertView, @NonNull ViewGroup
+                if (emojiResult != null) {
+                    Emoji emoji = emojiResult.emoji;
+                    String format = context.getString(R.string.emoji_shortcode_format);
+                    String formattedShortcode = String.format(format, emoji.getShortcode());
```
`context.getString` takes format arguments, you could save a line of code if you want

autumnontape commented:
All right, I've addressed all of the issues @connyduck brought up. I'm going to see if I can figure out how to put a horizontal rule in the autocomplete results now.

autumnontape commented:
Hmm... The default divider is exactly the same color as the autocomplete results' background in the dark theme.

autumnontape commented:
All right, I had to add some new resources to get the line to show up in the dark theme, but I added a divider. Is it good? Should I change/revert it?

![divider](https://user-images.githubusercontent.com/40726037/53677160-63c6a300-3c60-11e9-8102-a6b1a760849c.png)

autumnontape commented:
Oh, I might as well admit -- I have no idea what this `//noinspection ConstantConditions` comment I've been copy-pasting along with the layout inflating code in ComposeAutoCompleteAdapter.java actually does? So maybe some or all of the ones I added shouldn't be there.

charlag commented:
@autumnontape it's a marker for the IDE to not warn you about certain things, like it think that the variable is `@Nullable` but you know that it's not and don't want to check. Try to remove some and see if IDE complains.
You did a stellar work there :star:

autumnontape commented:
@charlag I removed all the "noinspection" comments in `ComposeAutoCompleteAdapter.getView()`, including the one that was already there, and ran a full build, but nothing seems to be complaining about it. So I guess either the original comment doesn't need to be there anymore, or whatever tool it was talking to isn't being run on my end/I'm not seeing its output.

autumnontape commented:
I am using Android Studio and not seeing anything in-editor either, FYI.

autumnontape commented:
Also FYI, I'm guessing my next push will be the last one, so it'll be another force-push to clean things up just a little.

autumnontape commented:
So all that's left for me is to possibly remove the `noinspection` comments since they don't seem to be doing anything. I have the "constant conditions and exceptions" inspection enabled, which I think is the one those comments are bypassing.

I just need to know if removing them is okay for you all or if I'm possibly missing something.

charlag commented:
Yes, please, I checked and it doesn't have a warning, feel free to remove them.

I had another look and it seems that shortcode format doesn't need to me translatable? Do I miss something here?
Otherwise I think it's good to go, asking @connyduck to merge it

autumnontape commented:
It wasn't marked `translatable="false"` because I copy-pasted it from this:

```xml
<string name="status_username_format">\@%s</string>
```

Which seemingly has identical "translations" across all the translation files.

In any case, I've now marked it as not translatable and removed the unnecessary `noinspection` comments, including the one that was there before, so I believe it's ready to merge :tada:

connyduck commented:
Awesome, thx!

autumnontape commented:
@connyduck Did you forget to merge this?

connyduck commented:
no, I did not want it in the 5.2 release
